drivers/r2dbc-driver/src/main/kotlin/app/cash/sqldelight/driver/r2dbc/R2dbcDriver.kt

```diff
@@ -201,27 +205,84 @@ class R2dbcPreparedStatement(private val statement: Statement) : SqlPreparedStat
   }
 }
 
-/**
- * TODO: Write a better async cursor API
- */
-class R2dbcCursor(val rowSet: List<List<Any?>>) : SqlCursor {
-  var row = -1
-    private set
+internal fun<T : Any> Publisher<T>.asIterator(): ChannelIterator<T> =
+  AsyncChannelIterator(this)
+
+private class AsyncChannelIterator<T : Any>(
+  pub: Publisher<T>,
+) : ChannelIterator<T> {
```

Comment:
Open to change the interface because we don't actually enforce the mentioned exceptions (ChannelClosedException).

Reply:
Decided to remove it because it makes implementation easier (no temp variables)